src/responser.rs

```diff
@@ -0,0 +1,684 @@
+        Self {
+            dpqn,
+            msn,
+            psn: last_psn,
+            last_retry_psn: None,
+        }
+    }
+
+    pub(crate) fn new_nack(dpqn: Qpn, msn: Msn, psn: Psn, last_retry_psn: Psn) -> Self {
+        Self {
+            dpqn,
+            msn,
+            psn,
+            last_retry_psn: Some(last_retry_psn),
+        }
+    }
+}
+
+/// Command about read response
+pub(crate) struct RespReadRespCommand {
+    pub(crate) desc: ToHostWorkRbDescRead,
+}
+
+/// The response command sent by other threads
+///
+/// Currently, it supports two types of response:
+/// * Acknowledge(ack,nack)
+/// * Read Response
+#[allow(dead_code)]
+pub(crate) enum RespCommand {
+    Acknowledge(RespAckCommand), // Acknowledge or Negative Acknowledge
+    ReadResponse(RespReadRespCommand),
+}
+
+/// A thread that is responsible for sending the response to the other side
+pub(crate) struct DescResponser {
+    _thread: std::thread::JoinHandle<()>,
+}
+
+impl DescResponser {
+    pub fn new(
+        device: Arc<dyn WorkDescriptorSender>,
+        recving_queue: std::sync::mpsc::Receiver<RespCommand>,
+        ack_buffers: AcknowledgeBuffer,
+        qp_table: Arc<RwLock<HashMap<Qpn, QpContext>>>,
+    ) -> Self {
+        let _thread = spawn(|| Self::working_thread(device, recving_queue, ack_buffers, qp_table));
+        Self { _thread }
+    }
+
+    fn working_thread(
+        device: Arc<dyn WorkDescriptorSender>,
+        recving_queue: std::sync::mpsc::Receiver<RespCommand>,
+        ack_buffers: AcknowledgeBuffer,
+        qp_table: Arc<RwLock<HashMap<Qpn, QpContext>>>,
+    ) {
+        loop {
+            match recving_queue.recv() {
+                Ok(RespCommand::Acknowledge(ack)) => {
+                    // send ack to device
+                    let ack_buf = ack_buffers.alloc().unwrap();
+                    let src_ip = Ipv4Addr::LOCALHOST;
+                    let (dst_ip, common) = match qp_table.read().unwrap().get(&ack.dpqn) {
+                        Some(qp) => {
+                            let dst_ip = qp.dqp_ip;
+                            let common = ToCardWorkRbDescCommon {
+                                total_len: ACKPACKET_SIZE as u32,
+                                rkey: Key::default(),
+                                raddr: 0,
+                                dqp_ip: qp.dqp_ip,
+                                dqpn: qp.qpn,
+                                mac_addr: qp.mac_addr,
+                                pmtu: qp.pmtu.clone(),
+                                flags: MemAccessTypeFlag::IbvAccessNoFlags,
+                                qp_type: QpType::RawPacket,
+                                psn: Psn::default(),
+                            };
+                            (dst_ip, common)
+                        }
+                        None => {
+                            eprintln!("Failed to get QP from QP table: {:?}", ack.dpqn);
```

Comment:
Can we replace this kind of `eprint` with `log::error()`?

Reply:
Yes. I have added the `clippy::print_stderr`, `clippy::print_stdout` lints and replaced them as log::error! etc.